be2-scala-dup/src/main/scala/ch/epfl/pop/decentralized/GossipManager.scala

```diff
@@ -0,0 +1,187 @@
+
+    // response is expected because only one entry exists
+    if (activeGossipPeers.size == 1) {
+      activeGossipPeers.foreach { (rumorRpc, _) =>
+        if (response.result.isEmpty && Random.nextDouble() < stopProbability) {
+          activeGossipProtocol -= rumorRpc
+        } else {
+          updateGossip(rumorRpc)
+        }
+      }
+    } else {
+      log.info(s"Unexpected match for active gossip. Response with id ${response.id} matched with ${activeGossipPeers.size} entries")
+      // removes duplicate entries to come back to a stable state
+      activeGossipPeers.foreach { (rumorRpc, _) =>
+        activeGossipProtocol -= rumorRpc
+      }
+    }
+  }
+
+  private def gossip(messages: Map[Channel, List[Message]]): Unit = {
+    val rumor: Rumor = Rumor(publicKey.get, rumorId, messages)
+    val jsonRpcRequest = prepareRumor(rumor)
+    rumorId += 1
+    updateGossip(jsonRpcRequest)
+  }
+
+  override def receive: Receive = {
+    case GossipManager.HandleRumor(jsonRpcRequest: JsonRpcRequest) =>
+      handleRumor(jsonRpcRequest)
+
+    case GossipManager.ManageGossipResponse(jsonRpcResponse) =>
+      processResponse(jsonRpcResponse)
+
+    case GossipManager.Gossip(messages) =>
+      gossip(messages)
+
+    case _ =>
+      log.info(s"Actor $self received an unexpected message")
+  }
+
+}
+
+object GossipManager extends AskPatternConstants {
+  def props(dbActorRef: AskableActorRef, monitorRef: ActorRef, connectionMediatorRef: AskableActorRef): Props =
+    Props(new GossipManager(dbActorRef, monitorRef, connectionMediatorRef))
+
+  def gossipHandler(gossipManager: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+    case Right(jsonRpcRequest: JsonRpcRequest) =>
+      jsonRpcRequest.method match
+        case MethodType.rumor =>
+          gossipManager ? HandleRumor(jsonRpcRequest)
+          Right(jsonRpcRequest)
+        case _ => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "GossipManager received a non expected jsonRpcRequest", jsonRpcRequest.id))
+    case graphMessage @ _ => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "GossipManager received an unexpected message:" + graphMessage, None))
+  }
+
+  def monitorResponse(gossipManager: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+    case Right(jsonRpcResponse: JsonRpcResponse) =>
+      gossipManager ? ManageGossipResponse(jsonRpcResponse)
+      Right(jsonRpcResponse)
+    case graphMessage @ _ => graphMessage
+  }
+
+  def gossip(gossipManager: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+    case Right(jsonRpcRequest: JsonRpcRequest) =>
+      jsonRpcRequest.getParamsMessage match
+        case Some(message) => gossipManager ? Gossip(Map(jsonRpcRequest.getParamsChannel -> List(message)))
+        case None          => /* Do nothing */
+      Right(jsonRpcRequest)
+    case graphMessage @ _ => graphMessage
```

Comment:
you can also add the case where it is a JsonRpcResponse and wrap it in a Left with pipelineError.